### PR TITLE
Start refactoring providers

### PR DESCRIPTION
**Pull request by MGaetan89**
While writing tests, I noticed that a lot of code in `sickbeard.providers` was duplicated. So I started refactoring that part to remove duplicated code and make the code more testable.

I will add more commits as I work on this and squash them all once I'm done, so you can easily see what's new.
If you have any remark/suggestion/feedback, don't wait until I'm done to share them with me ;)

### CODE REVIEW

**Comment by miigotu** [comment]
Ahhh I have been working on that slowly. Providers is the #1 point of failure so any change here has to be SUPER gradual.

**Comment by MGaetan89** [comment]
I'm hoping to change all the providers in this PR. That's why I won't squash the commits until the very end: so it's easier to review.
And with all the new unit tests that I will add, I hope it won't break anything (or at least not to much).
The logic itself is not going to change a lot. It will mostly be about the structure of the code.

**Comment by miigotu** [comment]
I have a lot of stuff that is not completed in there but ok. The stuff that seems duplicated in generic is not dupe code they are overrides.

**Comment by MGaetan89** [comment]
This PR is getting bigger than I thought, so I'll limit it to `GenericProvider`, `NZBProvider` and `TorrentProvider`. I'll make other PRs for the concrete providers.

Even if this PR is big, it does not change too much in the project. Here are the changes:
- `GenericProvider`, `NZBProvider` and `TorrentProvider` were moved from `sickbeard.providers` to `sickrage.providers` (in three separate files instead of one)
- Some methods were renamed to follow PEP8 recommendations
- Some methods have been added to simplify the code
- Some warnings have been fixed

Now that the build is passing again, I will write some tests to test those three new classes and then it will be ready to merge.

File: sickrage/providers/TorrentProvider.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    def _custom_trackers(self):
+        if sickbeard.TRACKERS_LIST:
+            if not self.public:
+                return ''
+
+            return '&tr=' + '&tr='.join(set([x.strip() for x in sickbeard.TRACKERS_LIST.split(',') if x.strip()]))
+
+        return ''
+
+    def _get_episode_search_strings(self, episode, add_string=''):
```
Comment (miigotu) [inline sickrage/providers/TorrentProvider.py]: these `_get_*_search_strings` methods need to be in GenericProvider, not just an abstract or some nzb providers wont work. Eventually they will all use this format.

There are some others that can be moved there as well. We want to try and override as little as possible.

Comment (MGaetan89) [inline sickrage/providers/TorrentProvider.py]: They are in `GenericProvider` https://github.com/SickRage/SickRage/pull/352/files#diff-7d601cb9cf37419c442e260e16686cb0R387
Those are just overriding the `GenericProvider` implementation.

Comment (miigotu) [inline sickrage/providers/TorrentProvider.py]: Im saying that these are the defaults that belong in GenericProvider, and dont need overridden in NZBProvider or TorrentProvider, as both use this format unless the actual provider overrides it directly.

Comment (MGaetan89) [inline sickrage/providers/TorrentProvider.py]: Ok, I understood it the other way, sorry.
I can move those to `GenericProvider`. This is just the way it is for the moment.

Comment (miigotu) [inline sickrage/providers/TorrentProvider.py]: The way it is now some NZB providers will not get search strings is why I pointed it out.

The end goal is to have as few overrides in NZBProvider, TorrentProvider, or the implementing classes, so anything that can be moved into the GenericProvider base class (without using any torrent or nzb specific imports or variables) and work for both should be to make for less code/overrides.

Comment (MGaetan89) [inline sickrage/providers/TorrentProvider.py]: I agree. I just didn't know that this was common code, as it is not in `NZBProvider`.
I will move that part to `GenericProvider` in my next commit.

**Comment by MGaetan89** [comment]
The last commits add tests to the three new classes (`GenericProvider`, `NZBProvider` and `TorrentProvider`).
It also moves the `_get_episode_search_strings` and `_get_season_search_strings` methods from `TorrentProvider` to `GenericProvider`, as suggested by @miigotu.

If it is good for you, I will squash the commits so you can merge this PR.

File: sickbeard/providers/bitsnoop.py [hunk 2]
```diff
@@ -47,7 +47,7 @@ def __init__(self):
 
         self.cache = BitSnoopCache(self)
 
-    def _doSearch(self, search_strings, search_mode='eponly', epcount=0, age=0, epObj=None): # pylint: disable=too-many-branches,too-many-arguments,too-many-locals
+    def _do_search(self, search_strings, search_mode='eponly', epcount=0, age=0, epObj=None): # pylint: disable=R0912,R0913,R0914
```
Comment (miigotu) [inline sickbeard/providers/bitsnoop.py:50]: use the short-message for pylint disables

**Comment by miigotu** [comment]
This can't be merged until at least the next merge to master. This also should have been done incrementally like I said at the beginning, because tests or not 4k changes in 76 files in 1 PR there will almost certainly be problems.

The classes should have been split up and moved in one PR and tested and merged, incubated for a week, rename methods in another PR and tested and merged and incubated for a week, and then start refactoring code. This puts a halt to the in-progress work i was doing also, some providers still even have findPropers overriden.

I cant even tell you how many times I have started doing this exact refactor and pulled out until it was more ready, and instead did smaller refactors to make the final one smaller.

This and the regexes are the most delicate parts of SR, and they require the most caution.

**Comment by miigotu** [comment]
I just released to master again, so I'll merge this so it can start being tested. Really please just try to keep the next ones smaller so everyone (me?) is more comfortable and confident is what they are merging. If the tower falls over I'm always the one it lands on lol.

**Comment by MGaetan89** [comment]
I understand your point. But most of this PR is moving the file to the `sickrage` package and rename the methods. Unless I missed something it should be good.
For the real providers implementation, I will do only 1 or 2 per PR. Now that the big changes are done, the following PRs are going to be much smaller.

**Comment by miigotu** [comment]
If you just get the first one or two moved into the sickrage module, I can work through the rest if you want. Moving them one by one will let me be able to see which I have done (which keeping track of was always an issue for me) and I will also be able to do the next stage of my incremental refactor at the same time where all result items are a dict with a preset structure.

**Comment by MGaetan89** [comment]
Yes, we can do it like that.
I am just not sure how to update `sickbeard/providers/__init__.py` so providers can be in two modules. I'll have to look more into that.
Or if you want to changes the providers first, we can postpone the remaining providers until then. It will give more time to test this PR's change.

**Comment by miigotu** [comment]
Well in `sickbeard/providers/__init__.py` you would just change the line that imports the provider itself and leave the `__all__` alone. Since it is lazy loaded everything else in SR would still find it.

When I was doing this/planning I wanted to use a subfolder for nzb and one for torrent, so that managing them and even adding more provider types in the future would feel more coherent.

I can make some time to start that tomorrow

**Comment by MGaetan89** [comment]
PyCharm collapses the import block, so I didn't notice that each provider is explicitly imported.

Ok, I can work on testing other part of SickRage (some that requires less changes :P) until you are ready for me to move the providers to `sickrage.providers`.

**Comment by miigotu** [comment]
Im just going to move them one by one to `sickrage.providers.` as I complete them

**Pull request merged**